flink-cdc-connect/flink-cdc-source-connectors/flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/reader/fetch/EventProcessorFactory.java

```diff
@@ -0,0 +1,282 @@
+                ChangeEventSource.ChangeEventSourceContext context,
+                OracleConnectorConfig connectorConfig,
+                OracleConnection jdbcConnection,
+                JdbcSourceEventDispatcher<OraclePartition> dispatcher,
+                OraclePartition partition,
+                OracleOffsetContext offsetContext,
+                OracleDatabaseSchema schema,
+                OracleStreamingChangeEventSourceMetrics metrics,
+                ErrorHandler errorHandler,
+                StreamSplit redoLogSplit) {
+            super(
+                    context,
+                    connectorConfig,
+                    jdbcConnection,
+                    dispatcher,
+                    partition,
+                    offsetContext,
+                    schema,
+                    metrics);
+            this.redoLogSplit = redoLogSplit;
+            this.errorHandler = errorHandler;
+            this.context = context;
+            this.dispatcher = dispatcher;
+        }
+
+        @Override
+        protected void processRow(OraclePartition partition, LogMinerEventRow row)
+                throws SQLException, InterruptedException {
+            super.processRow(partition, row);
+            afterRowProcess(partition, row, redoLogSplit, errorHandler, dispatcher, context);
+        }
+    }
+
+    public static void afterRowProcess(
+            OraclePartition partition,
+            LogMinerEventRow row,
+            StreamSplit redoLogSplit,
+            ErrorHandler errorHandler,
+            JdbcSourceEventDispatcher dispatcher,
+            ChangeEventSource.ChangeEventSourceContext context) {
+        // check do we need to stop for fetch redo log for snapshot split.
+        if (isBoundedRead(redoLogSplit)) {
+            final RedoLogOffset currentRedoLogOffset = new RedoLogOffset(row.getScn().longValue());
+            // reach the high watermark, the redo log fetcher should be finished
+            if (currentRedoLogOffset.isAtOrAfter(redoLogSplit.getEndingOffset())) {
+                // send redo log end event
+                try {
+                    dispatcher.dispatchWatermarkEvent(
+                            partition.getSourcePartition(),
+                            redoLogSplit,
+                            currentRedoLogOffset,
+                            WatermarkKind.END);
+                } catch (InterruptedException e) {
+                    LOG.error("Send signal event error.", e);
+                    errorHandler.setProducerThrowable(
+                            new DebeziumException("Error processing redo log signal event", e));
+                }
+                // tell fetcher the redo log task finished
+                ((OracleScanFetchTask.OracleSnapshotRedoLogSplitChangeEventSourceContext) context)
+                        .finished();
+            }
+        }
+    }
+
+    private static boolean isBoundedRead(StreamSplit redoLogSplit) {
```

Comment:
@loserwang1024 Can you move to utils and merge other same code.

Reply:
I'd like to do it, but I have no idea how to  merge other same code. It seems that each type of Offset have different NO_STOPPING_OFFSET:
* mysql: `BinlogOffset NO_STOPPING_OFFSET = new BinlogOffset("", Long.MIN_VALUE);`
* mongo: `ChangeStreamOffset NO_STOPPING_OFFSET = new ChangeStreamOffset(maximumBsonTimestamp());`
* sqlserver: `LsnOffset NO_STOPPING_OFFSET = new LsnOffset(Lsn.valueOf(new byte[] {Byte.MAX_VALUE}));`
* postgres: RedoLogOffset NO_STOPPING_OFFSET = new RedoLogOffset(Long.MIN_VALUE)
* oracle: `RedoLogOffset NO_STOPPING_OFFSET = new RedoLogOffset(Long.MIN_VALUE);`

Reply:
Maybe open another pr for this will be better.